### ISSUE TITLE
Object Relational Mapping exercise Task 4 has no hint

### ISSUE THREAD

**Issue by christian-elm** [exercism/csharp#1973]
For a beginner, a hint here would be helpful here. Manually managing resources is quite a daunting task when having not dealt with it before.

**Comment by christian-elm** [comment]
@aage No, not specifically a question on how to implement the interface. What tripped me up as a junior was that task 4 aborts a database transaction, but I as the "author" of the code don't have direct control of state - calling Database.Write changes the database to a state, but the method caller calls instance.Dispose() which also changes the state. So I would have liked a hint on how to work around this problem when not having direct control of the state of the data layer.

What I can see in the file isn't relevant enough for Task 4 to be duplicated.

**Comment by christian-elm** [comment]
Also just for consistency's sake it would be nice that all tasks have their hints (they are always very helpful when you get stuck), or at least the tasks that are required in order to unlock new concepts and exercises.

**Comment by aage** [comment]
@christian-elm I'm sorry, but I don't understand what is unclear. The instructions mention the `Orm` class needs to clean up the `Database` half way through a transaction (`Dispose` right after `Write`). The task is to implement this code.

Do you have a suggestion of what a hint might be that would clear this up? Adding hints to the two tasks that have none is easy enough.

**Comment by christian-elm** [comment]
@aage It might be a bit difficult for me as a novice to explain. What trips me up: in Task 2 I'm instructed to call `Database.Write(),` which changes the state of the database to `DataWritten` unless an exception is thrown. So, when the test method calls `orm.Dispose()`, the state of the database has already changed to `DataWritten`. 

My understanding of the exercise instructions is that this is expected behavior, so: what I fail to understand is how I can prevent the state of the database to change to `DataWritten` when I have implemented a `Database.Write()` call in task 2.

If this still doesn't make it clear, perhaps it's better to close this issue. Unfortunately I don't have any suggestions for hints as I have yet to complete the exercise myself, which means I don't yet have a good grip on how to properly implement the `IDisposable` interface.

**Comment by aage** [comment]
@christian-elm I don't think you're supposed to prevent the state being `DataWritten`, I think the whole point is _how_ you clean up when being in the middle of a transaction. Hint: you probably can just call `Dispose` on the `Database`. I hope this helps.

Can this issue be closed as far as you're concerned?

**Comment by aage** [comment]
@christian-elm I've given the code another look, I think I know now what you mean. When calling the database you should _Dispose_ the database when an `Exception` occurs. This goes for the `Write` method and all the others (except `Dispose`). The _bad write_ throws an exception which you can [try/catch](https://www.w3schools.com/cs/cs_exceptions.php).

Hopefully this answers your question, sorry for the confusion on my end!

**Comment by christian-elm** [comment]
@aage I managed to figure it out eventually and solve the exercise. Thanks for sticking with it and trying to help! You can probably close this issue now.

**Comment by aage** [comment]
@christian-elm Great!

@ErikSchierboom Can you close this issue?
